WIP: MGMT-299 and MGMT-368 documentation

abhatt-rh opened the pull request:
@ahardin-rh @mburke5678: Request you to kindly review the documentation updates for doc stories [TELCODOCS-75](https://issues.redhat.com/browse/TELCODOCS-75) and [TELCODOCS-33](https://issues.redhat.com/browse/TELCODOCS-33) 

Refers to engineering epics MGMT-299 and MGMT-368 for which the the corresponding doc epics are [TELCODOCS-142](https://issues.redhat.com/browse/TELCODOCS-142) and [TELCODOCS-188](https://issues.redhat.com/browse/TELCODOCS-188), respectively.

ahardin-rh commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,112 @@
+
+Instead of reprovisioning the nodes, power-based remediation uses a power controller to power off an inoperable node. This type of remediation is also called power fencing.
+
+The OpenShift Container Platform uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.
```
s/The OpenShift Container Platform/{product-title}

ahardin-rh commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,112 @@
+Usually bare metal reprovisioning is a lengthy process, during which the cluster
+is missing compute resources and applications might be interrupted.
+To change the default remediation process from machine deletion to host power-cycle,
+annotate the MachineHealthCheck resource with the
```
`MachineHealthCheck` resource 


per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

ahardin-rh commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,112 @@
+
+.Prerequisites
+
+* The OpenShift Container Platform is installed using installer-provisioned infrastructure (IPI).
```
s/The OpenShift Container Platform/{product-title}

ahardin-rh commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,112 @@
+. Create a `healthcheck.yml` file that contains the definition of your machine health check.
+. Apply the `healthcheck.yml` file to your cluster using the following command:
+
+[source,terminal]
```
[source,terminal]
    ----
    $ oc apply -f healthcheck.yml
    ----

ahardin-rh commented:
@abhatt-rh Is this for versions 4.6+?

ahardin-rh commented:
Just a few minor comments from me. Thanks!

mburke5678 commented:
@abhatt-rh The new structure looks good to me.

abhatt-rh commented:
@ahardin-rh and @mburke5678 : Thank you for your review. I have incorporated the comments.

@n1r1 @beekhof @shellymiron Request you to kindly review the changes for the Fencing epics here::https://deploy-preview-31697--osdocs.netlify.app/openshift-enterprise/latest/machine_management/deploying-machine-health-checks.html#mgmt-power-remediation-baremetal-about_deploying-machine-health-checks

abhatt-rh commented:
@aireilly Kindly review

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+The remediation process operates as follows:
+
+. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.
+. The MHC notifies the controller in the cluster API provider for managed bare metal hardware (CAPM3) controller.
```
in openshift we use CAPBM (not CAPM3)

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.
+. The MHC notifies the controller in the cluster API provider for managed bare metal hardware (CAPM3) controller.
+. The controller in the cluster API provider for managed bare metal hardware (CAPM3) requests to power-off the unhealthy node.
+. After the power is off, the Node CR is deleted, which allows the cluster to reschedule the affected workload on other nodes.
```
s/CR/resource as `node` is not a *custom* resource

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+. The controller in the cluster API provider for managed bare metal hardware (CAPM3) requests to power-off the unhealthy node.
+. After the power is off, the Node CR is deleted, which allows the cluster to reschedule the affected workload on other nodes.
+. The cluster API provider for managed bare metal hardware (CAPM3) requests to power on the node.
+. After the node is up, the node re-registers itself with the cluster, causing the creation of a new Node CR
```
same here for CR

beekhof replied:
I think that was my fault, but yes "Node resource" would be more technically correct

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+. After the power is off, the Node CR is deleted, which allows the cluster to reschedule the affected workload on other nodes.
+. The cluster API provider for managed bare metal hardware (CAPM3) requests to power on the node.
+. After the node is up, the node re-registers itself with the cluster, causing the creation of a new Node CR
+. After  the Node CR is recreated, the cluster API provider for managed bare metal hardware (CAPM3) restores the annotations and labels that existed on the unhealthy node before its deletion.
```
double space,

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+
+[NOTE]
+====
+If the power operations did not complete, the cluster API provider for managed bare metal hardware (CAPM3) triggers the reprovisioning of the unhealthy node.
```
unless this is a master node or node that can not be reprovisioned (e.g. `ExternallyProvisioned`)

beekhof commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+Power-based remediation meets the following objectives:
+
+* Allows the recovery of control plane nodes
+* Prevents data loss in hyperconverged environments.
```
If the node was left running, then in theory it could be generating or processing data that isn't being synchronised to the rest of the cluster.  Which would count as data loss.

Likewise if we recover a master, the cluster can survive the subsequent loss of another one - allowing incoming data to be process and not lost.

So technically true, but possibly a little misleading.

```suggestion
* Reduces the risk data loss in hyperconverged environments.
* Reduces the downtime associated with recovering physical machines
```

beekhof commented on modules/machine-health-checks-resource.adoc:
```diff
@@ -7,47 +7,7 @@
-====
-The `matchLabels` are examples only; you must map your machine groups based on your specific needs.
-====
+The `MachineHealthCheck` resource resembles the following YAML file:
```
Is there a link missing?

abhatt-rh replied:
No, Andrew. There is no link since the section talks about a sample YAML file that is displayed for MHC

beekhof commented on modules/machine-health-checks-resource.adoc:
```diff
@@ -7,47 +7,7 @@
-====
+The `MachineHealthCheck` resource resembles the following YAML file:
 
 .`MachineHealthCheck` for all other installation types
```
```suggestion
.`MachineHealthCheck` for all cloud based installations
```

beekhof commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+
+The {product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.
+
+Power-based remediation meets the following objectives:
```
```suggestion
Power-based remediation provides the following capabilities:
```

beekhof commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+is missing compute resources and applications might be interrupted.
+To change the default remediation process from machine deletion to host power-cycle,
+annotate the `MachineHealthCheck` resource with the
+`machine.openshift.io/remediation-strategy: external-baremetal` annotation.
```
Note that this part of the procedure will change in 4.9

abhatt-rh replied:
@aireilly There is a sample YAML file in the topic that shows the use of this annotation

aireilly commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+<2> For bare metal clusters, you must include the `machine.openshift.io/remediation-strategy: external-baremetal` annotation in the `annotations` section to enable power-cycle remediation. With this remediation strategy, unhealthy hosts are rebooted instead of removed from the cluster.
+<3> Specify a label for the machine pool that you want to check.
+<4> Specify the machine set to track in `<cluster_name>-<label>-<zone>` format. For example, `prod-node-us-east-1a`.
+<5> Specify the timeout duration for a node condition. If a condition is met for the duration of the timeout, the machine will be remediated. Long timeouts can result in long periods of downtime for a workload on an unhealthy machine.
```
```suggestion
<5> Specify the timeout duration for the node condition. If the condition is met for the duration of the timeout, the machine will be remediated. Long timeouts can result in long periods of downtime for a workload on an unhealthy machine.
```

abhatt-rh commented on modules/machine-health-checks-resource.adoc:
```diff
@@ -7,47 +7,7 @@
-====
+The `MachineHealthCheck` resource resembles the following YAML file:
 
 .`MachineHealthCheck` for all other installation types
```
Rephrasing to the following:
`MachineHealthCheck` for all installation types other than bare metal.

abhatt-rh commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,114 @@
+
+[source,terminal]
+----
+$ oc apply -f healthcheck.yml
```
Could not find any cue on the yml vs yaml, but since most instances in the docs use `.yaml` or YAML, so changed it to `healthcheck.yaml`

abhatt-rh commented:
@n1r1 @beekhof I have completed addressing the comments and we are ready to merge the PR. Which versions do we merge it to? I think it will be 4.6 and later. Kindly confirm.

shellymiron commented:
/lgtm

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,113 @@
+
+Instead of reprovisioning the nodes, power-based remediation uses a power controller to power off an inoperable node. This type of remediation is also called power fencing.
+
+{product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.
```
`MachineHealthCheck` controller doesn't detect faulty nodes by initiating power-cycling.

it detect faulty nodes in the same way for all platforms. power-cycling is the remediation action taken in BM.

abhatt-rh replied:
Rephrased to: 
{product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes. Power-based remediation is fast and reboots faulty nodes instead of removing them from the cluster.

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,113 @@
+The remediation process operates as follows:
+
+. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.
+. The MHC notifies the bare metal machine controller requests to power-off the unhealthy node.
```
Something is slightly off here?
maybe:
The MHC notifies the bare metal machine controller **which** requests to power-off the unhealthy node.

n1r1 commented on modules/mgmt-power-remediation-baremetal-about.adoc:
```diff
@@ -0,0 +1,113 @@
+
+[NOTE]
+====
+If the power operations did not complete, the bare metal machine controller triggers the reprovisioning of the unhealthy node unless this is a master node or a node that is reprovisioned externally.
```
```suggestion
If the power operations did not complete, the bare metal machine controller triggers the reprovisioning of the unhealthy node unless this is a master node or a node that was provisioned externally.
```

n1r1 reviewed (commented):
Looks good!
left some nits below.

n1r1 commented:
/lgtm

ahardin-rh commented:
/cherrypick enterprise-4.7

ahardin-rh commented:
/cherrypick enterprise-4.8
